File: typescript/src/choice.ts
export interface Choice<V = unknown> {
  name: string;
  value: V;
  disabled?: boolean | string;
  short?: string;
  description?: string;
}

export interface Separator {
  type: "separator";
  text: string;
}

export type ChoiceItem<V = unknown> = Choice<V> | Separator;
export type RawChoice<V = unknown> = string | Choice<V> | Separator;

export function isSeparator(item: ChoiceItem): item is Separator {
  return "type" in item && item.type === "separator";
}

/**
 * Type-aware value matching (R4). An answer matches a choice's value only if it
 * has the same JSON type and value. JS strict equality already refuses to
 * cross-match types (`"42" !== 42`, `true !== 1`, `0 !== false`), so we never
 * string-coerce. Arrays/objects are compared structurally via canonical JSON.
 */
export function valuesMatch(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  // Distinguish primitives strictly: only fall back to structural comparison
  // when both sides are non-null objects of the same kind.
  if (typeof a !== "object" || typeof b !== "object" || a === null || b === null) {
    return false;
  }
  if (Array.isArray(a) !== Array.isArray(b)) return false;
  try {
    return JSON.stringify(a) === JSON.stringify(b);
  } catch {
    return false;
  }
}

/**
 * Build the canonical "Invalid choice" validation message shared across the
 * select / checkbox / rawlist / expand prompts. The rejected answer and each
 * valid option are encoded as compact JSON, and the valid options are joined by
 * ", " (comma + single space). This string is byte-identical across all four
 * language implementations (conformance parity), e.g.
 *   Invalid choice: "rs". Valid: ["py", "go"]
 *   Invalid choice: 1.5. Valid: ["313", "311"]
 */
export function invalidChoiceMessage(answer: unknown, validValues: unknown[]): string {
  const valid = validValues.map((v) => JSON.stringify(v)).join(", ");
  return `Invalid choice: ${JSON.stringify(answer)}. Valid: [${valid}]`;
}

export function createSeparator(text = "────────"): Separator {
  return { type: "separator", text };
}

export function parseChoice<V = unknown>(raw: RawChoice<V>): ChoiceItem<V> {
  if (typeof raw === "string") {
    // When a raw string is passed, it serves as both name and value.
    // The caller is responsible for ensuring V is compatible with string
    // (e.g., RawChoice<string> or RawChoice<unknown>).
    return { name: raw, value: raw as V };
  }
  if ("type" in raw && raw.type === "separator") {
    return raw;
  }
  // A choice with a missing value defaults its value to its name (R4).
  if (!("value" in raw) || (raw as Choice<V>).value === undefined) {
    return { ...(raw as Choice<V>), value: (raw as Choice<V>).name as unknown as V };
  }
  return raw;
}

export function choiceToDict(item: ChoiceItem): Record<string, unknown> {
  if (isSeparator(item)) {
    return { type: "separator", text: item.text };
  }
  const d: Record<string, unknown> = { name: item.name, value: item.value };
  if (item.disabled) d.disabled = item.disabled;
  if (item.short != null) d.short = item.short;
  if (item.description != null) d.description = item.description;
  return d;
}

File: rust/src/choice.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Choice {
    pub name: String,
    pub value: Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub disabled: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub short: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

impl Choice {
    pub fn new(name: impl Into<String>, value: impl Into<Value>) -> Self {
        Self {
            name: name.into(),
            value: value.into(),
            disabled: None,
            short: None,
            description: None,
        }
    }

    pub fn is_disabled(&self) -> bool {
        match &self.disabled {
            None => false,
            Some(Value::Bool(b)) => *b,
            Some(Value::String(s)) => !s.is_empty(),
            _ => true,
        }
    }

    pub fn disabled_reason(&self) -> Option<&str> {
        match &self.disabled {
            Some(Value::String(s)) => Some(s),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Separator {
    #[serde(rename = "type")]
    pub kind: String,
    pub text: String,
}

impl Separator {
    pub fn new(text: impl Into<String>) -> Self {
        Self {
            kind: "separator".to_string(),
            text: text.into(),
        }
    }
}

impl Default for Separator {
    fn default() -> Self {
        Self::new("────────")
    }
}

#[derive(Debug, Clone)]
pub enum ChoiceItem {
    Choice(Choice),
    Separator(Separator),
}

impl ChoiceItem {
    pub fn to_json(&self) -> Value {
        match self {
            Self::Choice(c) => serde_json::to_value(c).unwrap_or(Value::Null),
            Self::Separator(s) => serde_json::to_value(s).unwrap_or(Value::Null),
        }
    }
}

pub fn parse_choice(raw: &str) -> ChoiceItem {
    ChoiceItem::Choice(Choice::new(raw, Value::String(raw.to_string())))
}

pub fn parse_choice_from_value(val: Value) -> ChoiceItem {
    match &val {
        Value::String(s) => ChoiceItem::Choice(Choice::new(s.as_str(), val.clone())),
        Value::Object(map) => {
            if map.get("type").and_then(|v| v.as_str()) == Some("separator") {
                let text = map
                    .get("text")
                    .and_then(|v| v.as_str())
                    .unwrap_or("────────");
                ChoiceItem::Separator(Separator::new(text))
            } else {
                // Build the choice field-by-field so that a missing `value`
                // defaults to `name` (rather than stringifying the whole
                // object, which would never match an answer).
                let name = match map.get("name") {
                    Some(Value::String(s)) => s.clone(),
                    Some(other) => other.to_string(),
                    None => match map.get("value") {
                        Some(Value::String(s)) => s.clone(),
                        Some(other) => other.to_string(),
                        None => String::new(),
                    },
                };
                let value = map
                    .get("value")
                    .cloned()
                    .unwrap_or_else(|| Value::String(name.clone()));
                ChoiceItem::Choice(Choice {
                    name,
                    value,
                    disabled: map.get("disabled").cloned(),
                    short: map.get("short").and_then(|v| v.as_str()).map(String::from),
                    description: map
                        .get("description")
                        .and_then(|v| v.as_str())
                        .map(String::from),
                })
            }
        }
        _ => ChoiceItem::Choice(Choice::new(val.to_string(), val)),
    }
}
